File: apps/desktop/src-tauri/src/chat/tool_router.rs
use slop_ai::{affordances_to_tools, format_tree, LlmFunction, LlmTool, SlopNode, ToolSet};

/// Context built from all connected providers in a workspace.
pub struct MergedContext {
    pub tools: Vec<LlmTool>,
    pub tool_sets: Vec<(String, ToolSet)>, // (provider_id, tool_set)
    pub state_str: String,
    pub single_provider: bool,
    pub provider_names: Vec<(String, String)>, // (provider_id, provider_name)
}

/// Build merged context from connected providers.
pub fn build_merged_context(
    providers: &[(String, String, SlopNode)], // (id, name, tree)
) -> MergedContext {
    let single_provider = providers.len() == 1;
    let mut all_tools: Vec<LlmTool> = Vec::new();
    let mut tool_sets: Vec<(String, ToolSet)> = Vec::new();
    let mut state_str = String::new();
    let mut provider_names = Vec::new();

    for (id, name, tree) in providers {
        provider_names.push((id.clone(), name.clone()));
        let root_path = format!("/{}", tree.id);
        let ts = affordances_to_tools(tree, &root_path);

        if single_provider {
            all_tools.extend(ts.tools.iter().cloned());
        } else {
            // Prefix each tool with provider name for disambiguation
            for tool in &ts.tools {
                all_tools.push(LlmTool {
                    tool_type: tool.tool_type.clone(),
                    function: LlmFunction {
                        name: format!("{}__{}", name, tool.function.name),
                        description: format!("[{}] {}", name, tool.function.description),
                        parameters: tool.function.parameters.clone(),
                    },
                });
            }
            state_str.push_str(&format!("\n--- {} ---\n", name));
        }

        tool_sets.push((id.clone(), ts));
        state_str.push_str(&format_tree(tree, 0));
        state_str.push('\n');
    }

    MergedContext {
        tools: all_tools,
        tool_sets,
        state_str,
        single_provider,
        provider_names,
    }
}

/// Route a tool call name to (provider_id, path, action).
pub fn route_tool_call(
    tool_name: &str,
    ctx: &MergedContext,
) -> Option<(String, String, String)> {
    if ctx.single_provider {
        if let Some((id, ts)) = ctx.tool_sets.first() {
            if let Some(res) = ts.resolve(tool_name) {
                return Some((id.clone(), res.path.clone(), res.action.clone()));
            }
        }
        return None;
    }

    // Multi-provider: strip prefix to find target
    for (id, name) in &ctx.provider_names {
        let prefix = format!("{name}__");
        if tool_name.starts_with(&prefix) {
            let original_name = &tool_name[prefix.len()..];
            // Find the tool_set for this provider
            if let Some((_, ts)) = ctx.tool_sets.iter().find(|(pid, _)| pid == id) {
                if let Some(res) = ts.resolve(original_name) {
                    return Some((id.clone(), res.path.clone(), res.action.clone()));
                }
            }
            return None;
        }
    }
    None
}

File: apps/extension/src/background/tool-router.ts
import type { SlopNode, LlmTool, ToolSet } from "@slop-ai/consumer/browser";
import { affordancesToTools, formatTree } from "@slop-ai/consumer/browser";

export interface MergedContext {
  tools: LlmTool[];
  stateStr: string;
  singleProvider: boolean;
  providerNames: Array<{ name: string; index: number }>;
  /** Resolve a tool name (possibly provider-prefixed) back to path + action. */
  resolve(toolName: string): { providerIndex: number; path: string | null; action: string; targets?: string[] } | null;
}

export interface ProviderTreeInfo {
  name: string;
  index: number;
  tree: SlopNode;
}

export function buildMergedContext(providers: ProviderTreeInfo[]): MergedContext {
  const singleProvider = providers.length === 1;
  const allTools: LlmTool[] = [];
  let stateStr = "";
  const providerNames = providers.map(p => ({ name: p.name, index: p.index }));

  // Per-provider ToolSets for resolve
  const providerToolSets: { name: string; index: number; toolSet: ToolSet }[] = [];

  for (const { name, index, tree } of providers) {
    const toolSet = affordancesToTools(tree);
    providerToolSets.push({ name, index, toolSet });

    if (singleProvider) {
      allTools.push(...toolSet.tools);
    } else {
      const safeName = name.replace(/[^a-zA-Z0-9]/g, "_");
      for (const tool of toolSet.tools) {
        allTools.push({
          ...tool,
          function: {
            ...tool.function,
            name: `${safeName}__${tool.function.name}`,
            description: `[${name}] ${tool.function.description}`,
          },
        });
      }
      stateStr += `\n--- ${name} ---\n`;
    }

    stateStr += formatTree(tree) + "\n";
  }

  return {
    tools: allTools,
    stateStr,
    singleProvider,
    providerNames,
    resolve(toolName: string) {
      if (singleProvider) {
        const entry = providerToolSets[0];
        if (!entry) return null;
        const resolved = entry.toolSet.resolve(toolName);
        if (!resolved) return null;
        return { providerIndex: entry.index, ...resolved };
      }

      // Multi-provider: strip provider prefix, then resolve
      for (const entry of providerToolSets) {
        const safeName = entry.name.replace(/[^a-zA-Z0-9]/g, "_");
        const prefix = `${safeName}__`;
        if (toolName.startsWith(prefix)) {
          const unprefixed = toolName.slice(prefix.length);
          const resolved = entry.toolSet.resolve(unprefixed);
          if (!resolved) return null;
          return { providerIndex: entry.index, ...resolved };
        }
      }
      return null;
    },
  };
}
